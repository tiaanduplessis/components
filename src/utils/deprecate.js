/* eslint-disable no-console */
import {useRef, useCallback} from 'react'

const noop = () => {}
// eslint-disable-next-line import/no-mutable-exports
let deprecate = null

if (__DEV__) {
  deprecate = ({name, message, version}) => {
    Deprecations.deprecate({name, message, version})
  }
} else {
  deprecate = noop
}

export {deprecate}

// eslint-disable-next-line import/no-mutable-exports
let useDeprecation = null

if (__DEV__) {
  useDeprecation = ({name, message, version}) => {
    const ref = useRef(false)
    const logDeprecation = useCallback(() => {
      if (!ref.current) {
        ref.current = true
        deprecate({name, message, version})
      }
    }, [name, message, version])

    return logDeprecation
  }
} else {
  useDeprecation = () => {
    return noop
  }
}

export {useDeprecation}

export class Deprecations {
  static get() {
    if (!Deprecations.instance) {
      Deprecations.instance = new Deprecations()
    }

    return Deprecations.instance
  }

  constructor() {
    this.deprecations = []
  }

  static deprecate({name, message, version}) {
    const msg = `WARNING! ${name} is deprecated and will be removed in version ${version}. ${message}`
    console.warn(msg)

    this.get().deprecations.push({name, message, version})
  }

  static getDeprecations() {
    return this.get().deprecations
  }

  static clearDeprecations() {
    this.get().deprecations.length = 0
  }
}
