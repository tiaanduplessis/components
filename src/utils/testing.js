/* eslint-disable jsdoc/require-param, jsdoc/require-returns, jsdoc/require-description-complete-sentence, jsdoc/match-description */
import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

export const COMPONENT_DISPLAY_NAME_REGEX = /^[A-Z][A-Za-z]+(\.[A-Z][A-Za-z]+)*$/

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * using react-test-renderer and return the root node
 * ```
 */
export function renderRoot(component) {
  return renderer.create(component).root
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function percent(value) {
  return typeof value === 'number' ? `${value}%` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    // const dest = computed[key] || (computed[key] = {})
    const dest = {}
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }

    // Don't add media rule to computed styles
    // if no styles were actually applied
    if (Object.keys(dest).length > 0) {
      computed[key] = dest
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return node.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}

// If a component requires certain props or other conditions in order
// to render without errors, you can pass a `toRender` function that
// returns an element ready to be rendered.
export function behavesAsComponent(Component, systemPropArray, toRender = null, options) {
  if (typeof toRender === 'object' && !options) {
    options = toRender
    toRender = null
  }
  options = options || {}

  const getElement = () => (toRender ? toRender() : <Component />)

  it('implements system props', () => {
    for (const systemProps of systemPropArray) {
      expect(Component).toImplementSystemProps(systemProps)
    }
  })

  if (!options.skipSx) {
    it('implements the sx prop', () => {
      expect(Component).toImplementSxProp()
    })

    it('implements sx prop behavior', () => {
      expect(getElement()).toImplementSxBehavior()
    })
  }

  if (!options.skipAs) {
    it('respects the as prop', () => {
      const As = React.forwardRef((_props, ref) => <div className="as-component" ref={ref} />)
      const elem = React.cloneElement(getElement(), {as: As})
      expect(render(elem)).toEqual(render(<As />))
    })
  }

  it('sets a valid displayName', () => {
    expect(Component.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
  })

  it('sets the default theme', () => {
    expect(getElement()).toSetDefaultTheme()
  })

  it('renders consistently', () => {
    expect(render(getElement())).toMatchSnapshot()
  })
}

export function checkExports(path, exports) {
  it('has declared exports', () => {
    const mod = require(`../${path}`)
    expect(mod).toSetExports(exports)
  })
}
