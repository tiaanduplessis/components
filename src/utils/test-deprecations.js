import semver from 'semver'
import {Deprecations} from '../utils/deprecate'

const ourVersion = require('../../package.json').version

beforeEach(() => {
  Deprecations.clearDeprecations()
})

afterEach(() => {
  const deprecations = Deprecations.getDeprecations()

  for (const dep of deprecations) {
    if (semver.gte(ourVersion, dep.version)) {
      throw new Error(`Found a deprecation that should be removed in ${dep.version}`)
    }
  }
})
