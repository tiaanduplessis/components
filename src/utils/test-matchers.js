import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount, getComputedStyles, render} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toImplementSxProp(Component) {
    return {
      pass: !!Component.propTypes.sx,
      message: () => 'Missing sx propTypes'
    }
  },

  toImplementSxBehavior(element) {
    const mediaKey = '@media (max-width:123px)'
    const sxPropValue = {
      [mediaKey]: {
        color: 'red.5'
      }
    }

    const elem = React.cloneElement(element, {sx: sxPropValue})
    const rendered = render(elem)

    function checkStylesDeep(rendered) {
      const className = rendered.props ? rendered.props.className : ''
      const styles = getComputedStyles(className)
      if (styles[mediaKey] && styles[mediaKey].color) {
        return true
      } else if (rendered.children) {
        return rendered.children.some(child => checkStylesDeep(child))
      } else {
        return false
      }
    }

    return {
      pass: checkStylesDeep(rendered),
      message: () => 'sx prop values did not change styles of component nor of any sub-components'
    }
  },

  toSetDefaultTheme(Component) {
    let comp
    if (Component.type) {
      comp = Component
    } else {
      comp = <Component />
    }
    const wrapper = mount(comp)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  },

  toSetExports(mod, expectedExports) {
    if (!Object.keys(expectedExports).includes('default')) {
      return {
        pass: false,
        message: () => "You must specify the module's default export"
      }
    }

    const seen = new Set()
    for (const exp of Object.keys(expectedExports)) {
      seen.add(exp)
      if (mod[exp] !== expectedExports[exp]) {
        if (!mod[exp] && !expectedExports[exp]) {
          continue
        }

        return {
          pass: false,
          message: () => `Module exported a different value from key '${exp}' than expected`
        }
      }
    }

    for (const exp of Object.keys(mod)) {
      if (seen.has(exp)) {
        continue
      }

      if (mod[exp] !== expectedExports[exp]) {
        return {
          pass: false,
          message: () => `Module exported an unexpected value from key '${exp}'`
        }
      }
    }

    return {
      pass: true,
      message: () => ''
    }
  }
})
