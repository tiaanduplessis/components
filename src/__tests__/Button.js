import React from 'react'
import {Button, ButtonPrimary, ButtonDanger, ButtonOutline, ButtonGroup, ButtonTableList} from '..'
import {render, behavesAsComponent, checkExports} from '../utils/testing'
import {COMMON, FLEX, LAYOUT, TYPOGRAPHY} from '../constants'
import {render as HTMLRender, cleanup} from '@testing-library/react'
import {axe, toHaveNoViolations} from 'jest-axe'
import 'babel-polyfill'
expect.extend(toHaveNoViolations)

function noop() {}

describe('Button', () => {
  behavesAsComponent(Button, [COMMON, LAYOUT])

  checkExports('Button', {
    default: Button,
    ButtonPrimary,
    ButtonDanger,
    ButtonOutline,
    ButtonGroup,
    ButtonTableList
  })

  it('renders a <button>', () => {
    expect(render(<Button />).type).toEqual('button')
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Button>Click here</Button>)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
    cleanup()
  })

  it('preserves "onClick" prop', () => {
    expect(render(<Button onClick={noop} />).props.onClick).toEqual(noop)
  })

  it('respects width props', () => {
    expect(render(<Button width={200} />)).toHaveStyleRule('width', '200px')
  })

  it('respects the "disabled" prop', () => {
    const item = render(<Button disabled />)
    expect(item.props.disabled).toEqual(true)
    expect(item).toMatchSnapshot()
  })

  it('respects the "variant" prop', () => {
    expect(render(<Button variant="small" />)).toHaveStyleRule('font-size', '12px')
    expect(render(<Button variant="large" />)).toHaveStyleRule('font-size', '16px')
  })

  it('respects the "fontSize" prop over the "variant" prop', () => {
    expect(render(<Button variant="small" fontSize={20} />)).toHaveStyleRule('font-size', '20px')
  })
})

describe('ButtonPrimary', () => {
  behavesAsComponent(ButtonPrimary, [COMMON, LAYOUT])

  it('renders a <button>', () => {
    expect(render(<ButtonPrimary />).type).toEqual('button')
  })

  it('renders correct disabled styles', () => {
    const item = render(<ButtonPrimary disabled />)
    expect(item).toMatchSnapshot()
  })
})

describe('ButtonDanger', () => {
  behavesAsComponent(ButtonDanger, [COMMON, LAYOUT])

  it('renders a <button>', () => {
    expect(render(<ButtonDanger />).type).toEqual('button')
  })

  it('renders correct disabled styles', () => {
    const item = render(<ButtonDanger disabled />)
    expect(item).toMatchSnapshot()
  })
})

describe('ButtonOutline', () => {
  behavesAsComponent(ButtonOutline, [COMMON, LAYOUT])

  it('renders a <button> by default', () => {
    expect(render(<ButtonOutline />).type).toEqual('button')
  })

  it('renders correct disabled styles', () => {
    const item = render(<ButtonOutline disabled />)
    expect(item).toMatchSnapshot()
  })
})

describe('ButtonGroup', () => {
  behavesAsComponent(ButtonGroup, [COMMON, FLEX, LAYOUT])
})

describe('ButtonTableList', () => {
  behavesAsComponent(ButtonTableList, [COMMON, TYPOGRAPHY, LAYOUT])
})
