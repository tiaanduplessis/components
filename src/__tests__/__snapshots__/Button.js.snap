// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders consistently 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #24292e;
  background-color: #fafbfc;
  border: 1px solid rgba(27,31,35,0.12);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.04),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  background-color: #F3F4F6;
  box-shadow: 0px 1px 0px rgba(209,213,218,0.2),inset 0px 2px 0px rgba(255,255,255,0.1);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(3,102,214,0.3);
}

.c0:active {
  background-color: #edeff2;
  box-shadow: inset 0px 2px 0px rgba(149,157,165,0.1);
  border-color: #d1d5da;
}

.c0:disabled {
  color: #959da5;
  background-color: #fafbfc;
  border-color: #eaecef;
}

<button
  className="c0"
/>
`;

exports[`Button respects the "disabled" prop 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #24292e;
  background-color: #fafbfc;
  border: 1px solid rgba(27,31,35,0.12);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.04),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  background-color: #F3F4F6;
  box-shadow: 0px 1px 0px rgba(209,213,218,0.2),inset 0px 2px 0px rgba(255,255,255,0.1);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(3,102,214,0.3);
}

.c0:active {
  background-color: #edeff2;
  box-shadow: inset 0px 2px 0px rgba(149,157,165,0.1);
  border-color: #d1d5da;
}

.c0:disabled {
  color: #959da5;
  background-color: #fafbfc;
  border-color: #eaecef;
}

<button
  className="c0"
  disabled={true}
/>
`;

exports[`ButtonDanger renders consistently 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #cb2431;
  border: 1px solid #e1e4e8;
  background-color: #fafbfc;
  box-shadow: 0px 1px 0px rgba(149,157,165,0.1),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  color: #fff;
  background-color: #cb2431;
  border-color: rgba(27,31,35,0.15);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(203,36,49,0.4);
}

.c0:active {
  color: #fff;
  background-color: #be222e;
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
  border-color: rgba(27,31,35,0.15);
}

.c0:disabled {
  color: rgba(203,36,49,.5);
  background-color: #F3F4F6;
  border: 1px solid #e1e4e8;
}

<button
  className="c0"
/>
`;

exports[`ButtonDanger renders correct disabled styles 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #cb2431;
  border: 1px solid #e1e4e8;
  background-color: #fafbfc;
  box-shadow: 0px 1px 0px rgba(149,157,165,0.1),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  color: #fff;
  background-color: #cb2431;
  border-color: rgba(27,31,35,0.15);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(203,36,49,0.4);
}

.c0:active {
  color: #fff;
  background-color: #be222e;
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
  border-color: rgba(27,31,35,0.15);
}

.c0:disabled {
  color: rgba(203,36,49,.5);
  background-color: #F3F4F6;
  border: 1px solid #e1e4e8;
}

<button
  className="c0"
  disabled={true}
/>
`;

exports[`ButtonGroup renders consistently 1`] = `
.c0 {
  display: inline-block;
  display: inline-block;
  vertical-align: middle;
}

.c0.c0 > * {
  position: relative;
  border-right-width: 0;
  border-radius: 0;
}

.c0.c0 > *:first-child {
  border-top-left-radius: 6px;
  border-bottom-left-radius: 6px;
  margin-right: 0;
}

.c0.c0 > *:not(:first-child) {
  margin-left: 0;
  margin-right: 0;
}

.c0.c0 > *:last-child {
  border-right-width: 1px;
  border-top-right-radius: 6px;
  border-bottom-right-radius: 6px;
}

.c0.c0 > *:focus,
.c0.c0 > *:active,
.c0.c0 > *:hover {
  border-right-width: 1px;
}

.c0.c0 > *:focus + *,
.c0.c0 > *:active + *,
.c0.c0 > *:hover + * {
  border-left-width: 0;
}

.c0.c0 > *:focus,
.c0.c0 > *:active {
  z-index: 1;
}

<div
  className="c0"
  display="inline-block"
/>
`;

exports[`ButtonOutline renders consistently 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #0366d6;
  border: 1px solid #e1e4e8;
  background-color: #fafbfc;
  box-shadow: 0px 1px 0px rgba(149,157,165,0.1),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  color: #fff;
  background-color: #0366d6;
  border-color: rgba(27,31,35,0.15);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(3,102,214,0.3);
}

.c0:active {
  color: #fff;
  background-color: #035fc7;
  border-color: rgba(27,31,35,0.15);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:disabled {
  color: #959da5;
  border-color: #e1e4e8;
  background-color: #F3F4F6;
}

<button
  className="c0"
/>
`;

exports[`ButtonOutline renders correct disabled styles 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #0366d6;
  border: 1px solid #e1e4e8;
  background-color: #fafbfc;
  box-shadow: 0px 1px 0px rgba(149,157,165,0.1),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  color: #fff;
  background-color: #0366d6;
  border-color: rgba(27,31,35,0.15);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(3,102,214,0.3);
}

.c0:active {
  color: #fff;
  background-color: #035fc7;
  border-color: rgba(27,31,35,0.15);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:disabled {
  color: #959da5;
  border-color: #e1e4e8;
  background-color: #F3F4F6;
}

<button
  className="c0"
  disabled={true}
/>
`;

exports[`ButtonPrimary renders consistently 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #fff;
  background-color: #2EA44F;
  border: 1px solid #22863a;
  box-shadow: 0px 1px 0px rgba(20,70,32,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  background-color: #2C974B;
  border-color: rgba(27,31,35,0.15);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px #94D3A2;
  background-color: #2C974B;
}

.c0:active {
  background-color: #128031;
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
  border-color: rgba(27,31,35,0.15);
}

.c0:disabled {
  color: rgba(255,255,255,0.50);
  background-color: #94D3A2;
  border-color: rgba(34,134,58,0.1);
}

<button
  className="c0"
/>
`;

exports[`ButtonPrimary renders correct disabled styles 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #fff;
  background-color: #2EA44F;
  border: 1px solid #22863a;
  box-shadow: 0px 1px 0px rgba(20,70,32,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus {
  outline: none;
}

.c0:disabled {
  cursor: default;
}

.c0:disabled svg {
  opacity: 0.6;
}

.c0:hover {
  background-color: #2C974B;
  border-color: rgba(27,31,35,0.15);
}

.c0:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px #94D3A2;
  background-color: #2C974B;
}

.c0:active {
  background-color: #128031;
  box-shadow: 0px 1px 0px rgba(27,31,35,0.1),inset 0px 2px 0px rgba(255,255,255,0.03);
  border-color: rgba(27,31,35,0.15);
}

.c0:disabled {
  color: rgba(255,255,255,0.50);
  background-color: #94D3A2;
  border-color: rgba(34,134,58,0.1);
}

<button
  className="c0"
  disabled={true}
/>
`;

exports[`ButtonTableList renders consistently 1`] = `
.c0 {
  display: inline-block;
  padding: 0;
  font-size: 14px;
  color: #586069;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:disabled,
.c0:disabled:hover {
  color: rgba(#586069,0.5);
  cursor: default;
}

.c0:after {
  display: inline-block;
  margin-left: 4px;
  width: 0;
  height: 0;
  vertical-align: -2px;
  content: "";
  border: 4px solid transparent;
  border-top-color: currentcolor;
}

<summary
  className="c0"
/>
`;
